paddle/operators/get_places_op.cc

```diff
@@ -0,0 +1,78 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/framework/op_registry.h"
+#include "paddle/platform/place.h"
+#ifdef PADDLE_WITH_CUDA
+#include "paddle/platform/gpu_info.h"
+#endif
+
+namespace paddle {
+namespace operators {
+
+class GetPlacesOp : public framework::OperatorBase {
+ public:
+  GetPlacesOp(const std::string &type, const framework::VariableNameMap &inputs,
+              const framework::VariableNameMap &outputs,
+              const framework::AttributeMap &attrs)
+      : OperatorBase(type, inputs, outputs, attrs) {}
+  void Run(const framework::Scope &scope,
+           const platform::DeviceContext &dev_ctx) const override {
+    std::string device_type = Attr<std::string>("device_type");
+    auto device_count = Attr<int>("device_count");
+
+    auto out_var_name = Output("Out");
+    auto *out_var = scope.FindVar(out_var_name);
+    PADDLE_ENFORCE(out_var != nullptr, "Output variable %s cannot be found",
+                   out_var_name);
+
+    auto &places = *(out_var->GetMutable<std::vector<platform::Place>>());
+    places.resize(device_count);
+    if (device_type == "CUDA") {
+#ifdef PADDLE_WITH_CUDA
+      PADDLE_ENFORCE_LT(device_count, platform::GetCUDADeviceCount());
+      for (int i = 0; i < device_count; i++) {
+        places.emplace_back(platform::GPUPlace(i));
+      }
+#else
+      PADDLE_THROW("'GPUPlace' is not supported in CPU only device.");
+#endif
+    } else if (device_type == "CPU") {
```

Comment:
What is the relationship between device_count and  CPU cores？Or do we need to limit the device_count to a ratio large than CPU cores?

Reply:
I'm confusing why to specify a ratio number, but I think the `device_count` means the CPU cores when `device_type="CPU"`.

Reply:
If device_count = 0, then it will use all devices of CPUs or GPUs.